pkg/sql/logictest/tests/cockroach-go-testserver-23.2/cockroach-go-testserver-23_2_test: TestLogic_mixed_version_bootstrap_tenant failed

pkg/sql/logictest/tests/cockroach-go-testserver-23.2/cockroach-go-testserver-23_2_test.TestLogic_mixed_version_bootstrap_tenant [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17213358?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17213358?buildTab=artifacts#/) on release-24.2 @ [bb3c7255bfe22efa15308add80fd8958cfc0071c](https://github.com/cockroachdb/cockroach/commits/bb3c7255bfe22efa15308add80fd8958cfc0071c):


```
        io.copyBuffer({0xffff35b27898, 0x4001ca81b0}, {0x6b82b00, 0x400207e050}, {0x0, 0x0, 0x0})
        	GOROOT/src/io/io.go:429 +0x18c
        io.Copy(...)
        	GOROOT/src/io/io.go:388
        os.genericWriteTo(0x4003fc2080?, {0xffff35b27898, 0x4001ca81b0})
        	GOROOT/src/os/file.go:269 +0x5c
        os.(*File).WriteTo(0x4003fc2080, {0xffff35b27898, 0x4001ca81b0})
        	GOROOT/src/os/file.go:247 +0xa0
        io.copyBuffer({0xffff35b27898, 0x4001ca81b0}, {0x6b829a0, 0x4003fc2080}, {0x0, 0x0, 0x0})
        	GOROOT/src/io/io.go:411 +0x98
        io.Copy(...)
        	GOROOT/src/io/io.go:388
        os/exec.(*Cmd).writerDescriptor.func1()
        	GOROOT/src/os/exec/exec.go:578 +0x44
        os/exec.(*Cmd).Start.func2(0x0?)
        	GOROOT/src/os/exec/exec.go:728 +0x34
        created by os/exec.(*Cmd).Start in goroutine 27
        	GOROOT/src/os/exec/exec.go:727 +0x7e0
        Leaked goroutine: goroutine 86 [IO wait]:
        internal/poll.runtime_pollWait(0xffff8062d120, 0x72)
        	GOROOT/src/runtime/netpoll.go:345 +0xa0
        internal/poll.(*pollDesc).wait(0x4003fc4780?, 0x4003f38000?, 0x1)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x28
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0x4003fc4780, {0x4003f38000, 0x8000, 0x8000})
        	GOROOT/src/internal/poll/fd_unix.go:164 +0x200
        os.(*File).read(...)
        	GOROOT/src/os/file_posix.go:29
        os.(*File).Read(0x4003fc2098, {0x4003f38000?, 0xd1?, 0xffff806f4a68?})
        	GOROOT/src/os/file.go:118 +0x70
        io.copyBuffer({0xffff35b27898, 0x4001ca81c8}, {0x6b82b00, 0x4000248768}, {0x0, 0x0, 0x0})
        	GOROOT/src/io/io.go:429 +0x18c
        io.Copy(...)
        	GOROOT/src/io/io.go:388
        os.genericWriteTo(0x4003fc2098?, {0xffff35b27898, 0x4001ca81c8})
        	GOROOT/src/os/file.go:269 +0x5c
        os.(*File).WriteTo(0x4003fc2098, {0xffff35b27898, 0x4001ca81c8})
        	GOROOT/src/os/file.go:247 +0xa0
        io.copyBuffer({0xffff35b27898, 0x4001ca81c8}, {0x6b829a0, 0x4003fc2098}, {0x0, 0x0, 0x0})
        	GOROOT/src/io/io.go:411 +0x98
        io.Copy(...)
        	GOROOT/src/io/io.go:388
        os/exec.(*Cmd).writerDescriptor.func1()
        	GOROOT/src/os/exec/exec.go:578 +0x44
        os/exec.(*Cmd).Start.func2(0x0?)
        	GOROOT/src/os/exec/exec.go:728 +0x34
        created by os/exec.(*Cmd).Start in goroutine 27
        	GOROOT/src/os/exec/exec.go:727 +0x7e0
--- FAIL: TestLogic_mixed_version_bootstrap_tenant (67.92s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_mixed_version_bootstrap_tenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-42841

--- a/pkg/sql/logictest/logic.go
+++ b/pkg/sql/logictest/logic.go
@@ -1352,7 +1352,7 @@ func (t *logicTest) newTestServerCluster(bootstrapBinaryPath, upgradeBinaryPath 
 
 	ts, err := testserver.NewTestServer(opts...)
 	if err != nil {
-		t.handleWaitForInitErr(err)
+		t.handleWaitForInitErr(ts, err)
 	}
 	t.testserverCluster = ts
 	t.clusterCleanupFuncs = append(t.clusterCleanupFuncs, ts.Stop, cleanupLogsDir)
@@ -1377,7 +1377,7 @@ func (t *logicTest) waitForAllNodes() {
 	for i := 0; i < t.cfg.NumNodes; i++ {
 		// Wait for each node to be reachable.
 		if err := t.testserverCluster.WaitForInitFinishForNode(i); err != nil {
-			t.handleWaitForInitErr(err)
+			t.handleWaitForInitErr(t.testserverCluster, err)
 		}
 	}
 }
@@ -1389,7 +1389,7 @@ func (t *logicTest) waitForAllNodes() {
 // issue, and we haven't been able to investigate it effectively, we
 // will ignore this error.
 // See https://github.com/cockroachdb/cockroach/issues/128759.
-func (t *logicTest) handleWaitForInitErr(err error) {
+func (t *logicTest) handleWaitForInitErr(ts testserver.TestServer, err error) {
 	if testutils.IsError(err, "init did not finish for node") {
 		foundSnappyErr := false
 		walkErr := filepath.WalkDir(t.logsDir, func(path string, d fs.DirEntry, err error) error {
@@ -1420,6 +1420,7 @@ func (t *logicTest) handleWaitForInitErr(err error) {
 		if walkErr != nil {
 			t.t().Logf("error while walking logs directory: %v", walkErr)
 		} else if foundSnappyErr {
+			ts.Stop()
 			t.t().Skip("ignoring init did not finish for node error due to snappy error")
 		}
 	}

